src/main/java/org/mskcc/cmo/messaging/impl/JSGatewayImpl.java

```diff
@@ -0,0 +1,180 @@
+package org.mskcc.cmo.messaging.impl;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.nats.client.Connection;
+import io.nats.client.Dispatcher;
+import io.nats.client.JetStream;
+import io.nats.client.JetStreamSubscription;
+import io.nats.client.Message;
+import io.nats.client.Nats;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.mskcc.cmo.common.FileUtil;
+import org.mskcc.cmo.common.impl.FileUtilImpl;
+import org.mskcc.cmo.messaging.Gateway;
+import org.mskcc.cmo.messaging.MessageConsumer;
+import org.mskcc.cmo.messaging.events.NATSPublisher;
+import org.mskcc.cmo.messaging.events.PublishingQueueTask;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JSGatewayImpl implements Gateway {
+    private static final Log LOG = LogFactory.getLog(JSGatewayImpl.class);
+    private static final CountDownLatch PUBLISHING_SHUTDOWN_LATCH = new CountDownLatch(1);
+    private static volatile Boolean shutdownInitiated = Boolean.FALSE;
+    private static volatile BlockingQueue<PublishingQueueTask> PUBLISHING_QUEUE = new LinkedBlockingQueue<>();
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final Map<String, JetStreamSubscription> subscribers = new HashMap<>();
+    private volatile ExecutorService exec = Executors.newSingleThreadExecutor();
+
+    // connection variables
+    @Value("${nats.url}")
+    public String natsUrl;
+
+    private Connection natsConnection;
+    private JetStream jsConnection;
+
+    // publishing logger file variables
+    @Value("${metadb.publishing_failures_filepath}")
+    private String metadbPubFailuresFilepath;
+
+    private final String PUB_FAILURES_FILE_HEADER = "DATE\tTOPIC\tMESSAGE\n";
+    private static final FileUtil fileUtil = new FileUtilImpl();
+    private static File publishingLoggerFile;
+
+    @Override
+    public void connect() throws Exception {
+        connect(natsUrl);
+    }
+
+    @Override
+    public void connect(String natsUrl) throws Exception {
+        this.natsConnection = Nats.connect(natsUrl);
```

Comment:
Unless this is being done somewhere else, we have to setup secure/tls connection.  I think we can reuse some of the  code that does this here:
[https://github.com/mskcc/cmo-messaging-java/blob/master/src/main/java/org/mskcc/cmo/messaging/impl/NATSGatewayImpl.java#L170](https://github.com/mskcc/cmo-messaging-java/blob/master/src/main/java/org/mskcc/cmo/messaging/impl/NATSGatewayImpl.java#L170)

Reply:
@divyamadala30 can you check this also?